Projekt styring (#99)
chhoumann/p7: report/sections/Chapter 11 - Reflection/Project-management.tex

In the following section we will describe how the project was managed, which techniques were used, as well as which issues we encountered. Throughout we will reflect on the decisions that were made in this context.

\section{Task management}

> One of the primary objectives in any project is to manage the tasks that are generated throughout the project. The goal is to have the ability to maintain an overview of what has been done, what needs to be done and how the tasks are distributed between developers. Additionally, project managers also need often want to be able to estimate how long tasks may take to complete and their difficulty. To accomplish this, projects may utilize several project management tools and techniques. In this section we will focus on the tools that we utilized, while in section \ref{sec:agile-dev} we will focus on the techniques that were utilized to manage the project.

Reviewer: ```suggestion
One of the primary objectives in any project is to manage the tasks that are generated throughout the project. The goal is to maintain an overview of what has been done, what needs to be done and how the tasks are distributed between developers. Additionally, project managers also need often want to be able to estimate how long tasks may take to complete and their difficulty. To accomplish this, projects may utilize several project management tools and techniques. In this section we will focus on the tools that we utilized, while in section \ref{sec:agile-dev} we will focus on the techniques that were utilized to manage the project.
```

\subsection{Github projects}
Initially we decided on using a tool called Github Projects. Github Projects is a project management tool that integrates issues, which is a text description of bugs and features that should be included in the project as well as pull requests into a board similar to a Kanban Board. This meant that we could create a board that was categorized into four different types: \textit{Todo}, \textit{In progress}, \textit{Ready for review} and \textit{Ready}, giving us a clear overview of what needed to be done, what was being worked on, what needed to be reviewed and what was done. All accessed from the same website. 

> While this tool offers a lot of benefits for developers in terms of project management, the tool was unfamiliar to all of the group members, as no member had any experience with the tool. The lack of experience became a hurdle within a few weeks of the project starting. This was due to several factors, including an unwillingness to understand the tool, but the primary issue that we were also learning other tools, frameworks and even a new programming language. All of which were more important to the project and therefore learning Github Projects became a less of a priority. To the point that we completely stopped using it, despite its promises.

Reviewer: Probably not great to write that we were unwilling. Rather, something like `the learning curve added friction to adopting the tool, which added extraneous overhead to existing developer habits. This prevented group members from properly using the tool, and therefore made it a net negative utility.`

Reviewer: This also supports the below sentence, about us reverting to tools we have experience with.

Reviewer: Would also like to see a deeper explanation of what part of GitHub Projects added the extra overhead.
The descriptions made the tools sound almost identical, so perhaps we can clarify a bit?

Here are my own thoughts on the matter:
- GH projects would be easier if we were more accustomed to working with issues in our git commits, as much of the automation comes from referencing issues in commits.
- However, when you create a PR without doing all of this, you have to manually assign things. And a PR could be a subtask of a 'story task', which is what our task board has. That made it a bit more annoying.
- Managing a project generally has more 'fluffy' tasks than what can be represented only in the development domain. We often have tasks unrelated to code, which is hard to track with GH projects.

Reviewer: ```
GitHub Projects seemed promising, as it integrates with existing developer workflows. You track issues, and as you work on them, they flow from the backlog to in progress, to in review, and then is at last categorized as completed. When you submit a pull request, it goes to the review section, so other members are aware that it is pending. And when that pull request is merged, it goes to completed automatically.
Intuitively, it makes sense to track issues like this, as it is so closely aligned with the modern Git workflow on GitHub. 
However, we found that some members were not accustomed to this way of working, and as such, didn't take the extra steps to facilitate the automation, e.g. marking issues as related to the project.
And when you create a pull request without referencing issues related to the project, you have to manually assign things later. Given that a pull request doesn't always correlate directly with issues, this is easily forgotten.
Managing a project generally includes managing non-technical tasks as well, which is hard to track with Projects, as it is very development-specific.
The final straw was bugs where the automation workflows weren't invoked, resulting in confusion. This resulted in us not using the tool.
```

This led us to revert back to tools that we had previously used.

\subsubsection{Notion}
Notion is a platform that allows users organize resources and manage projects in the same space. The platform allows you to customize the layout of your work space in a way that best suits the team. One particular feature of Notion that we used in this project was the Kanban board. In the Kanban board we defined four categories, similar to what we had in Github Projects, in which we could place work items:

\begin{itemize}

>     \item Not started.
>     \item In progress.
>     \item Review required.
>     \item Done.

Reviewer: ```suggestion
    \item Not started
    \item In progress
    \item Review required
    \item Done
```
Remove periods at end of items?

\end{itemize}

While Notion provides fewer features in terms of integration with Github, the tool was familiar to us and provided the functionality that we needed to manage the project.
Once we switched to Notion, planning and updating the tasks became a more active part of our daily work. This led to better productivity as we were now able to better maintain an overview of what each member of the group was doing and and how far they were with each task.

\section{Agile development approach} \label{sec:agile-dev}
Irrespective of which tools we used, the goal from the beginning of the project was to use a SCRUM-like approach to project management. 

> Generally, SCRUM is a framework that helps teams and organizations generate value through flexible solutions to complex problems. SCRUM is an iterative and incremental approach to project work where the project is broken up into sprints. A sprint is timeframe, usually 1-4 weeks in length, in which the the teams tries to complete the work that has been selected from a product backlog.

Reviewer: citation needed

> In the spirit of the SCRUM mindset we only adopted the specific things that we felt were necessary. For our project we did not have a SCRUM master or product owner. Instead we collectively created tasks based on features that we wanted to implement. 

Reviewer: ```suggestion
We only adopted the specific things from Scrum that we felt were necessary. For our project we did not have a Scrum master or product owner. Instead, we collectively created tasks based on features that we deem necessary to implement.
In traditional Scrum, you would have product owners that talk to customers about which features are important to them\todo{cite Scrum by Jeff Sutherland}. Since we are both stakeholders and developers, we found that acting as a council is more effective than delegating the role to a single member.
```
Reasoning:
- can't say 'we wanted to'... it's a bit weak of a reason, and we have a better one
- explaining our process for prioritization is important IMO. It's a huge part of project management.

> Additionally, instead of traditional sprints we decided on a minimally viable product, a product that would satisfy the project goals without any extra features, and divided that into sub parts. We would then generate tasks that would amount to the completion of that subtask and set a date for the expected completion. While this approach in essence is similar to a sprint, we did not think of it as a sprint. 

Reviewer: Shouldn't we explain why we opted to aim for MVP tasks, rather than using sprints?

Reviewer: We designed the specifications for a minimum viable product, consisting of only the core features we deemed necessary for the success of this type of project. This is akin to filling the project backlog with all features that could be in the project.
Then we went on to prioritize the features again, focusing on features that fulfilled the use-cases we imagined for the platform, grouping them as seen in the MOSCOW analysis in section \ref{...}\todo{fix ref}.
We found that sprints don't fit the nature of semester projects, as you have varying time to work, and it is usually less than half of the work week. We have found that smaller tasks and less analytical overhead, which Scrum brings, fits these projects well.

We also did not assign any points to the tasks in order to estimate the time needed to complete tasks. Instead these were just taken as a dicussion each morning, where we would brief each other on the estimated time remaining to complete the task.
Finally, we also approached review and retrospect differently in that we would handle potential process changes as issues arose. An example of this being the decision to switch to Notion. Similarly, discussions about internal team issues were taken regularly.

> This approach was based on previous experience working with a more rigid SCRUM methodology in a school project setting which we had found to be less productive. We felt that a more fluid approach to SCRUM worked the best especially due to the limited amount of work hours available dedicated specifically to the project per week.

Reviewer: Fine catch-all explanation, but maybe put it earlier, so the reader doesn't think what I did (see previous comments about explaining decisions).

> \todo{ting vi maaske kan udvide med: CI/CD, code review, pair programming - er der andet?}

Reviewer: Yes

\section{Task management}
One of the primary objectives in any project is to manage the tasks that are generated throughout the project. The goal is to maintain an overview of what has been done, what needs to be done and how the tasks are distributed between developers. Additionally, project managers also often want to be able to estimate how long tasks may take to complete and their difficulty. To accomplish this, projects may utilize several project management tools and techniques. In this section we will focus on the tools that we utilized, while in section \ref{sec:agile-dev} we will focus on the techniques that were utilized to manage the project.

\subsection{Github projects}
GitHub Projects seemed promising, as it integrates with existing developer workflows. You track issues, and as you work on them, they flow from the backlog to \textit{in progress}, to \textit{in review}, and then is at last categorized as \textit{completed}. When you submit a pull request, it goes to the review section, so other members are aware that it is pending. And when that pull request is merged, it goes to completed automatically.

> Intuitively, it makes sense to track issues like this, as it is so closely aligned with the modern Git workflow on GitHub. 

Reviewer: ```suggestion
Intuitively, it makes sense to track issues like this, as it is so closely aligned with the modern Git workflow on GitHub. To start contributing to a codebase, you create a branch. Then you reference issues when you make a pull request, which will then become linked to your changes, making issues closely tied to actual code changes.
```

Reviewer: Just to explain _what_ the modern workflow usually is

However, we found that some members were not accustomed to this way of working, and as such, did not take the extra steps to facilitate the automation, such as marking issues as related to the project.
When you create a pull request without referencing issues related to the project, you have to manually assign things later. Given that a pull request does not always correlate directly with issues, this is easily forgotten.
Managing a project generally includes managing non-technical tasks as well, which is hard to track with Projects, as it is very development-specific.

> What finally led us to consider other tools were bugs related to the automation workflows not being invoked, resulting in confusion.

Reviewer: ```suggestion
What finally led us to consider other tools were bugs related to the automated workflows not being invoked, resulting in confusion.
```

\section{Agile development approach} \label{sec:agile-dev}
Irrespective of which tools we used, the goal from the beginning of the project was to use a Scrum-like approach to project management.
The decision to this approach was based on previous experience working with a more rigid Scrum methodology in a school project setting which we had found to be less productive.
Generally, Scrum is a framework that helps teams and organizations generate value through flexible solutions to complex problems. Scrum is an iterative and incremental approach to project work where the project is broken up into sprints. A sprint being a time frame, usually 1-4 weeks in length, in which the the teams tries to complete the work that has been selected from a product backlog.

We only adopted the specific things from Scrum that we felt were necessary. For our project we did not have a Scrum master or product owner. Instead, we collectively created tasks based on features that we felt were necessary to implement.

> In traditional Scrum, you would have product owners that talk to customers about which features are important to them. Since all group members are both stakeholders and developers, we found that acting as a council is more effective than delegating the role to a single member.

Reviewer: This is where I'd put the cite. Not after saying what we do (l. 37)
